### PR TITLE
added capability to query the sim directly, named immediate query

### PR DESCRIPTION
**Pull request by armando-fandango**
### Proposed change(s)

Currently, the way the protocol buffers in the mlagents_env is designed, we can not send immediate queries to sim and get responses. For a project we had to add this feature, sending it as a pull request.

In the current setup, even if you add side channels, the response does not arrive unless the request is pushed before step() or reset() and response is received after one of these.

### Types of change(s)

- [ ] Bug fix
- [x] New feature [#5478](https://github.com/Unity-Technologies/ml-agents/issues/5478)
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/main/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/main/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/main/docs/Migrating.md) (if applicable)

### CODE REVIEW

**Comment by armando-fandango** [comment]
sorry it seems like I forgot to bring some files over from the changed version. will open this later.

**Comment by armando-fandango** [comment]
added all the modified files.

**Review by dongruoping — COMMENTED** [review]
Thanks for your contribution and apologies for the delayed review! I left a few comments below and overall it looks good. There are still a few things need to be added before we can merge the PR:
1. You need to add a public API for sending immediate messages in `environment.py`. The function should be public, with a clear name of its purpose, and with docstrings.
2. Please add some unit tests. The related tests are located under `ml-agents-envs/mlagents_envs/tests/test_envs.py`.

File: com.unity.ml-agents/Runtime/Communicator/RpcCommunicator.cs [hunk 1]
```diff
@@ -390,6 +391,72 @@ void SendBatchedMessageHelper()
 
             UpdateEnvironmentWithInput(rlInput);
 
+            foreach (var brainName in rlInput.AgentActions.Keys)
```
[the remark this answers is not in the document]
Comment (armando-fandango) [reply com.unity.ml-agents/Runtime/Communicator/RpcCommunicator.cs:394]: sorry for that, the commented lines have been removed and will be pushed with next push.

**Comment by armando-fandango** [comment]
> Thanks for your contribution and apologies for the delayed review! I left a few comments below and overall it looks good. There are still a few things need to be added before we can merge the PR:
> 
> 1. You need to add a public API for sending immediate messages in `environment.py`. The function should be public, with a clear name of its purpose, and with docstrings.

The process_immediate_message is now renamed to be a public function along with docstrings, would be pushing it with the next push in the next few days.

> 2. Please add some unit tests. The related tests are located under `ml-agents-envs/mlagents_envs/tests/test_envs.py`.

For adding a unit test, a mock environment is needed that can receive a query and return a response. Need some time to implement and hence would be pushing it with the next push in the next few days.

**Comment by miguelalonsojr** [comment]
Closing PR as stale.

**Pull request closed without merging**